[IMP] website: introduce cookies bar on website

rdeodoo opened the pull request:
With this commit, it is now possible to activate a cookies bar on a website
through the res.settings.

The cookies bar can be edited in the frontend by entering edit mode (to be
editable, it will be shown even if the user has hidden the cookies bar).

The cookies bar will appear until the user clicks on 'I agree'.

task-2087003

rdeodoo commented:
This one is missing the Popup mode of the cookies bar.
Also, it will be blocked until we finish the task about bottom left panel in web editor to show unvisible DOM element options (popup etc) to see if we can merge some feature of popup and this PR.

rdeodoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1066,6 +1070,75 @@ options.registry.anchor = options.Class.extend({
+            onSuccess: _widget => widget = _widget,
+        });
+
+        // TODO (review) need to activate 2 options, better way to do it?
```
@qsm-odoo

JKE-be commented:
Déjà discuté plusieurs fois, mais est ce qu el'on garde ca finalement ? 

Ou on part du principe que le user peut se debrouiller avec le s_popup (une fois le delay between show, montré)

Au derniere nouvelle, FP a l'air de ne plus vouloir dupliquer des snippets qui font la meme chose...
Eg: tab / pricing / multi pricing (= tab + pricing)

rdeodoo commented:
C'est complètement différent ici, soit on garde ça soit j'utilise pas popup imo.
Je sais pas si tu as testé, mais le left panel est tout à fait différent, grosso modo ça utilise surtout le code JS du widget au lieu de tout dupliquer, ce qui est bien :+1: 

Donc oui c'est considéré comme terminé cette tache à part l'imp de QSM que j'attends

JKE-be commented:
Apparement je n'ai plus rien a dire, on attend plus que le imp et le r+ de qsm, mais pour le log:

1. "Period Between Consents" n'est pas générique, ca devrait plutot être dans le sens "entre 2 display" et le plus court possible. Demander à sbu peut etre le mieux ?  mais dans le cas de news

2. Le warning de Fabien on select est imo too much, même si c'est une FP request.

3. Au uncheck de l'option, le warning ne devrait pas s'afficher à nouveau.

4. Un settings de plus dans le backend, quand on sait que tous les 2 mois on a une taches pour retirer des settings et en avoir le moins possible ... 

5. L'option position devrait être caché (utile au onchange pour preconfig des layout, mais peu utile pour le end user et permet de facilement arriver à un truc moche alors qu'on suggere 3 layouts de base qui marche bien + la size pour custo.)

 ps : et merci pour le imp du delai entre 2 affichages sur le popup snippet, nice !!

rdeodoo commented:
> Apparement je n'ai plus rien a dire, on attend plus que le imp et le r+ de qsm, mais pour le log:

Apparement je dois rester chez moi si j'ai pas envie de travailler, mais comme j'ai envie de travailler :

> 1. "Period Between Consents" n'est pas générique, ca devrait plutot être dans le sens "entre 2 display" et le plus court possible. Demander à sbu peut etre le mieux ?  mais dans le cas de news

C'est SBU qui a déjà trouvé ça, je le laisse revenir avec une meilleur idée @sbuhl 

> 2. Le warning de Fabien on select est imo too much, même si c'est une FP request.

J'ai essayé et réessayé TBH, pas réussi à lui faire retirer. Pour l'instant on a cookies mais avec ce compromis. Peut etre si quelqu'un d'autre le travail encore un peu.. Aux dernières nouvelles il doit revenir avec une meilleur phrase et un lien vers un article..

> 3. Au uncheck de l'option, le warning ne devrait pas s'afficher à nouveau.

My bad, oublié de push le diff, ça marche c'est déjà corrigé, je viens de push

> 4. Un settings de plus dans le backend, quand on sait que tous les 2 mois on a une taches pour retirer des settings et en avoir le moins possible ...

Ca va être retiré et mis dans le nouveau TAB où vont se trouver les customize show dés que possible (pas encore mergé le nouveau tab)

> 5. L'option position devrait être caché (utile au onchange pour preconfig des layout, mais peu utile pour le end user et permet de facilement arriver à un truc moche alors qu'on suggere 3 layouts de base qui marche bien + la size pour custo.)

Ok with that, je vais cacher l'option, je pense qu'ils ont assez avec les préset. Ca donnait une possibilité en plus mais pas beaucoup de value si on a les préset.

> ps : et merci pour le imp du delai entre 2 affichages sur le popup snippet, nice !!

:+1:

rdeodoo commented on addons/website/views/snippets/s_popup.xml:
```diff
@@ -40,23 +41,26 @@
             <we-colorpicker string="Backdrop Color" data-select-style="true"
                 data-css-property="background-color" data-color-prefix="bg-" data-dependencies="position_centered"/>
+        </div>
+        <div data-js="SnippetPopup" data-selector=".s_popup" data-exclude="#website_cookies_bar" data-target=".s_popup_main">
```
I feel like this should be done in JS with compute visibility, as I will need to also hide 'Position' which is the first one.
If I keep doing it this way, we will have to split again the data-js. I wrote it like that at first but I think it was not working since it is not its own snippet option (splitted data-JS).

Will investigate..

Plus not very modular, if another whatever code need to use this and hide other widget options (most likely wont happen), it won't be able to do it in XML anyway.

qsm-odoo replied:
Not sure to understand tell me if it is still an up-to-date issue ?

rdeodoo commented on addons/website/views/website_templates.xml:
```diff
@@ -271,6 +271,28 @@
+    <xpath expr="//div[@id='footer']" position="before">
+        <div id="website_cookies_bar" t-if="website.cookies_bar" class="s_popup o_snippet_invisible" data-name="Cookies Bar">
+            <div class='s_popup_main s_popup_fixed popup_backdrop d-none o_cookies_discrete'
+                 data-show-after="500" data-display="afterDelay" data-consents-delay="999">
```
@JKE-be Now I remember about the Python/JS Qweb template sharing, this template is a copy paste of one of the preview template we have in JS Qweb. Not really a problem but a bit of copy/pasted code to maintain

sbuhl commented on addons/website/static/src/js/backend/res_config_settings.js:
```diff
@@ -33,4 +39,49 @@ BaseSettingController.include({
+            )
+        ).append(_.str.sprintf(
+            _t("For session cookies, authentification and analytics, <b>you do not need to ask for the consent</b>, provided that your analytics is anonymized, which should be the case if you use Google Analytics. (%s, adopted in June 2012)"),
+            "<a target='_blank' href='https://ec.europa.eu/justice/article-29/documentation/opinion-recommendation/files/2012/wp194_en.pdf'>cf wp194 of Art 29</a>"
```
Approved by FP? Same for btns?

rdeodoo replied:
Done by FP actually, yeah..

odony replied:
Last time I checked, GA was not anonymous by default, [you need to opt-out of the storage of the last byte of the IP](https://support.google.com/analytics/answer/2763052?hl=en) - it's worth mentioning, isn't it?

Also why link to the PDF that could 404 at some point, rather than simply quote "*see e.g. Opinion 04/2012 on Cookie Consent Exemption by the EU Art.29 WP*", which will be easy to google for anyone who cares.

If we keep the harcoded link, making it "nofollow noopener noreferrer" makes sense I think, because it's external.

rdeodoo replied:
I'd rather have no popup on cookies feature activation tbh.
But you are right about the PDF link, unless we host it on our own server, we can't ensure it will be available forever.

odony replied:
Unless I missed something, current text still seems to imply that GA is anonymous by default?

rdeodoo replied:
Didn't process that part yet, should be good now

sbuhl commented on addons/website/views/snippets/s_popup.xml:
```diff
@@ -40,23 +41,26 @@
                 <we-button data-select-data-attribute="mouseExit">On Exit</we-button>
             </we-select>
             <we-input class="o_we_inline" string="Delay" data-select-data-attribute="" data-attribute-name="showAfter" data-unit="s" data-save-unit="ms" data-dependencies="show_delay"/>
+            <we-input class="o_we_inline" string="Period between consents" data-select-data-attribute="7days" data-attribute-name="consentsDelay" data-unit="days"/>
```
Alternatives:
_ Consent Duration
_ Expires after

rdeodoo replied:
`Consent Duration` IMO

rdeodoo replied:
As discussed in website office, we will go with `Hide For` for now and a tooltip.
Note that title are broken now, need https://github.com/odoo/odoo/pull/43111 in order to work

odony commented on addons/website/static/src/js/backend/res_config_settings.js:
```diff
@@ -33,4 +39,49 @@ BaseSettingController.include({
+            role: 'alert',
+        }).html(
+            $('<p/>').html(
+                _t("<b>Cookie bars are bad for the experience</b> of your visitors. We believe you should not annoy your visitors with such a practice, unless you absolutly need it.")
```
```suggestion
                _t("<b>Cookie bars may significantly impair the experience</b> of your visitors, especially on mobile devices. We recommend you avoid them unless you have verified with a legal advisor that you absolutely need cookie consent in your country.")
```

odony commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1027,6 +1030,61 @@ options.registry.anchor = options.Class.extend({
+        this.textButton = _t('I agree');
+        this.textTitle = _t('Respecting your privacy is our priority.');
+        this.textPrimary = _t('We use cookies to provide you a better user experience.');
+        this.textSecondary = _t('We use them to store info about your habits on our website. It will helps us to provide you the very best experience and customize what you see.') + '<br />' + _t('By clicking on this banner or navigating our website, you give us permission to collect data.');
```
"or navigating our website" is not a valid consent according to GDPR, since the user may not have see the banner at all. 

For the record, the conditions for a consent to be valid are strict and listed in recital 32 of GDPR: https://gdpr-info.eu/recitals/no-32/

Here we are also missing a link to the Cookie Policy or Privacy Policy with cookie info.

odony replied:
Also why are we harcoding so much text in JS, instead of putting it in a template. It's an anti-pattern ;)

rdeodoo replied:
> Also why are we harcoding so much text in JS, instead of putting it in a template. It's an anti-pattern ;) It would be easy to have the default values in the template.

Actually, this all started with the 'I agree' button which handles the cookies bar closing (and set cookies etc).
Since the template is fully editable, the user can easily remove the 'I agree' button (by mistake or not). Thus, the goal here is to make sure there is a way to restore that button. Otherwise, the feature would be broken forever as no button to hide the cookies bar.

At the same time, I also added 3 different templates covering the more common Cookie Bars style.
This was the opportunity to use that layout option to reset the button on template change.

Then, I extended the behavior to the Cookies Bar title, main text, second text, so the user can always get back to the template default layout, without losing what he typed (also true when changing from one layout to another). For instance, the user customize layout 1 with its own text, then switch to layout 2, we don't want to override what he just wrote but integrate it in the new layout.

If there is a better way to store default value directly inside the templates, without duplicating the code, I'd be able to use it but in this case all 3 templates use the same text by default, and the same text to restore a remove element during layout change option.

odony replied:
> We just discussed this with FP right now, using the Odoo.com Cookie Policy looks like a good compromise?
> At least, the user would have a complete feature ready to work, with a bootstraped Cookie Policy link/page.
> He will be able to edit it if he wants to.
> 
> What do you think?

Why not, hopefully they will get the hint and edit it according to their own use case. I guess it's enough to make it look like they are compliant, even if they don't know what they are doing.

rdeodoo replied:
Yes I think so too.
Obviously, as you say, they don't know what they are doing, but they think they do.
People are 100% sure cookies bar are mandatory. They don't mind it or overthink it. Every 'figure d'autorité' even has cookies bars (gov website, police, banks..).
So yeah the feature is just to fit customer needs. They do need it, at least they think they need it, just gotta check number of ticket about it.

rdeodoo commented:
> Why don't we let users directly edit the DOM after dropping a snippet with the default text? 

That's the point, he is able to fully edit it, meaning also breaking the template (eg by removing the I agree button).
We want to keep what he wrote when he changes the Cookie Bars layout.

We could probably move the text from JS to XML but that would be 3 times duplicated text. Not a fan but let me know what you think

odony commented on addons/website/views/snippets/s_popup.xml:
```diff
@@ -40,23 +41,26 @@
                 <we-button data-select-data-attribute="mouseExit">On Exit</we-button>
             </we-select>
             <we-input class="o_we_inline" string="Delay" data-select-data-attribute="" data-attribute-name="showAfter" data-unit="s" data-save-unit="ms" data-dependencies="show_delay"/>
+            <we-input class="o_we_inline" string="Hide For" title="Once the user close the popup, it won't be shown again for that period of time." data-select-data-attribute="7days" data-attribute-name="consentsDelay" data-unit="days"/>
```
```suggestion
            <we-input class="o_we_inline" string="Hide For" title="Once the user closes the popup, it won't be shown again for that period of time." data-select-data-attribute="7days" data-attribute-name="consentsDelay" data-unit="days"/>
```

odony commented on addons/website/static/src/js/backend/res_config_settings.js:
```diff
@@ -33,4 +39,49 @@ BaseSettingController.include({
+                click: cancelCallback,
+            },
+            {
+                text: 'Activate bad experience',
```
That button label sounds a bit childish to me. I'd rather not have the popup text at all either, but as we have it, why not make it more pro, e.g. "Activate anyway", or "Activate it despite the downsides" ?

rdeodoo replied:
That one if a difficult one TBH, we don't want it either but FP insisted to have it exactly named that way.
I am kinda stuck between your review and FP's one. It has been more than a year battle to be allowed to add a cookies bar in Odoo, but it came with some compromises from my/our part, mainly this modal on res setting.

I will check with FP for `Activate it despite the downsides` as you suggest. No point to add it now and change it back to `Activate bad experience` once merged when he sees it hehe

odony commented on addons/website/models/website.py:
```diff
@@ -168,6 +169,22 @@ def write(self, values):
+                if cookies_view and not cookies_page:
+                    self.env['website.page'].create({
+                        'is_published': True,
+                        'is_indexed': False,
```
Currently crashes with "no such field", making it harder to test the feature ;-)
```suggestion
                        'website_indexed': False,
```

odony commented:
> That's the point, he is able to fully edit it, meaning also breaking the template (eg by removing the I agree button).
> We want to keep what he wrote when he changes the Cookie Bars layout.

Tbh, I thought snippets were meant to be instantiated with drag-and-drop. So if people broke it they could just scrap it and drop the snippet again (or cancel before saving). Not sure why this one is enabled via a setting instead, and thus harder to reset.

Also, don't we generally use pure CSS on the container to customize snippets l&f (e.g. resetting the container classes or something), without touching the content? I guess it was not workable here.


> We could probably move the text from JS to XML but that would be 3 times duplicated text. Not a fan but let me know what you think

If that's an option to avoid more `t-raw`, I wouldn't mind! (that is, if there is no pure-CSS alternative, of course). Could make the templates more self-contained and readable too, with less code, no?

But it's not my spec, it's up to you. Just stating for the record that it looks like we're casually playing with XSS-prone things again, and the reasons are not very convincing. Things that initially looked safe can turn out not to be, especially after two refactorings, so it's best to avoid it as much as possible ;-)

qsm-odoo commented:
> Tbh, I thought snippets were meant to be instantiated with drag-and-drop. So if people broke it they could just scrap it and drop the snippet again (or cancel before saving). Not sure why this one is enabled via a setting instead, and thus harder to reset.

We try to keep it that way. FP wanted some snippets to be enabled thanks to a checkbox (like a popup on an unique page)... for now we only gave up for the cookie bar. I think it does actually make sense for this one since it is something that you want on every page and only once.

> Also, don't we generally use pure CSS on the container to customize snippets l&f (e.g. resetting the container classes or something), without touching the content? I guess it was not workable here.

Most of the time of course but we will get more and more snippets whose entire layout could change via options so it will not be about toggling classes or style anymore (actually that's not new, the gallery snippet which is there since 8.0 is about changing the whole DOM on option change). But we of course try to keep it about classes toggling when possible. I did not check this particular layout change in details yet but I guess there won't be any "CSS solution" (for me CSS is never a solution compared to clean DOM though...).

> > We could probably move the text from JS to XML but that would be 3 times duplicated text. Not a fan but let me know what you think
> 
> If that's an option to avoid more `t-raw`, I wouldn't mind! (that is, if there is no pure-CSS alternative, of course). Could make the templates more self-contained and readable too, with less code, no?

The default text should definitely be in the XML but that won't change the t-raw if I understand the code well. I am not sure about the problem, in the end it is just about moving existing DOM to a different location (if that existing DOM was a problem before... what side effect could be induced by moving it ?). Would you prefer a real move performed in JS instead of `.html()` -> `t-raw`? I mean, the template could contain no variable, use always the default text but when we re-render, we move the existing user text container node at the right location in the newly rendered template? It is the same in the end but maybe less prone-to-error-by-refactoring like you said?

qsm-odoo commented:
Also I forgot: FP accepted the "Activate anyway", I just asked him now :tada:

odony commented:
> We try to keep it that way. FP wanted some snippets to be enabled thanks to a checkbox (like a popup on an unique page)... for now we only gave up for the cookie bar. I think it does actually make sense for this one since it is something that you want on every page and only once.

Either way, the result is confusing. It behaves a bit like a snippet except you can't drag-and-drop it again, so it's brittle (people _will_ find a way to break it beyond repair, whatever you do, so more tickets).
And because it's activated somewhere else, users can't easily see that it is customisable or how, you just hope they notice it at some point.
If that's just a compromise for getting the feature in, fine, but if it's setting a new trend, it does not look like we're improving the UX.


> Would you prefer a real move performed in JS instead of `.html()` -> `t-raw`? I mean, the template could contain no variable, use always the default text but when we re-render, we move the existing user text container node at the right location in the newly rendered template? It is the same in the end but maybe less prone-to-error-by-refactoring like you said?

Yes, that sounds better :) There will be less transformations between DOM elements and JS strings (code vs data), which should be avoided as much as possible.
Every time we use `$.html()` or `t-raw` we should ask ourselves how we can avoid the conversion. In many cases the answer drives us to a solution that is much better. Result: template definitions really stay where everyone expect them (no parts of the text or DOM buried in JS), logic is less mixed with presentation, code is safer, more readable, etc.

qsm-odoo commented:
> so it's brittle (people _will_ find a way to break it beyond repair, whatever you do, so more tickets).
> And because it's activated somewhere else, users can't easily see that it is customisable or how, you just hope they notice it at some point.

Those checkboxes are being moved in the editor left panel in a new tab in one of my task, hopefully for 13.2. And once it is done we could simply reset the template entirely if it is disabled/re-enabled at some point. So I guess it will be fine at that moment.

> Result: template definitions really stay where everyone expect them (no parts of the text or DOM buried in JS), logic is less mixed with presentation, code is safer, more readable, etc.

Agreed :+1:

rdeodoo commented:
> Tbh, I thought snippets were meant to be instantiated with drag-and-drop. So if people broke it they could just scrap it and drop the snippet again (or cancel before saving).

To complete what @qsm-odoo said, cookies bar has more in common with the header (and probably footer) than real snippets. Not sure if already merged but you can change the header and layout through the left panel, and those are not snippet and not drag&dropable, just customization through the left panel.

> If that's an option to avoid more t-raw, I wouldn't mind! (that is, if there is no pure-CSS alternative, of course).

Removed the `t-raw`, removed the templating in JS. Everything is now inside XML templates.
Actually, even without taking `t-raw` into consideration, this is way better. Should've be that way at first, thanks for your input :+1: 

Also, everything discussed here (and not here :smile:) should've been implemented now.

odony commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1027,6 +1030,61 @@ options.registry.anchor = options.Class.extend({
+        ];
+
+        if (this.text === undefined) {
+            this.text = [];
```
More nitpicking (sorry!): storing markup in `text` variable is a bit confusing, like mixing `$.html()` and `$.text()`. Maybe name it `contentNodes`, or `contentSource`?

BTW, how about having all the content sections present in all 3 layouts, but some of them `d-none` somewhere? Consequences:
- No need to store the contents in JS vars at all, you can directly move the nodes to the new place with `$.appendTo()`. Bonus: no need for extracting `innerHTML` and reinjecting it, so we save one more `$.html` :tada: 
- The hidden contents is preserved for future layout changes as well, in another editor session, rather than lost after save.
- Minor drawback: the user might drop the hidden elements accidentally, so that data won't be there at the next layout change. No big deal, there are other ways to lose contents, including visible stuff. The transfer of contents between layouts is just a best effort thing.

odony replied:
On second thought, we end up polluting the DOM with hidden contents that may stay there forever, and that probably sucks. So nevermind, the above. Yet the variable name could still be improved ;-)

qsm-odoo replied:
Yep, we had some debate about this and I was against keeping d-none content in the DOM *unless it was customized by the user at some point and that it makes sense for the user not to lose it*. Our only example for now is our new "restaurant menu"-like snippet which may or not display descriptions for the menu items. On save, we keep the hidden descriptions **that were not customized**, so that the user may re-enable them someday but we remove the ones that were not customized to not pollute the DOM for no reason. Another potential reason, even if probably not the case anymore (the internet is not very clear about it): hidden text could be considered for SEO.

Anyways, I did not check the final result here yet. I'll see if we can do better yet.

odony commented:
> Actually, even without taking `t-raw` into consideration, this is way better. Should've be that way at first, thanks for your input 
> 
> Also, everything discussed here (and not here ) should've been implemented now.

Excellent, thanks a lot!

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1027,6 +1030,61 @@ options.registry.anchor = options.Class.extend({
+        }));
+
+        const $content = this.$target.find('.s_popup_content');
+        const selectorToKeep = [
```
selector**s** :)

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1027,6 +1030,61 @@ options.registry.anchor = options.Class.extend({
+                });
+            }
+            template.find(selector).html(this.text[selector]);
+        }
```
As @odony suggested indeed: instead of `innerHTML` -> `html`, use `childNodes` -> `appendChild` (wrong reason but worth mentioning: it should be 100x faster since using html requires parsing and recreation of DOM)

```javascript
this.allChildNodes = _.map($el, el => el.childNodes);
...
$template.find(selector).append(this.allChildNodes[selector]);
```

But in fact I am not sure it will be working since your use of `small` and `p` and ... etc. For example I wonder that if the user hits enter, it will create a new `p` which is probably not autorized in a `small` element and so might be extracted by smart browsers like Firefox ;) 

I wonder if each portion that you wanna keep can not be gather in a unique DIV then the code here would just be about moving that whole `div`? Instead of
```xml
<t t-name="default_stuff">Some text</t>
<small class="your_unique_selector">
    <t t-call="default_stuff"/>
</small>
```
Doing:
```xml
<t t-name="default_stuff"><div class="your_unique_selector">Some text</div></t>
<div class="small">
    <t t-call="default_stuff"/>
</div>
```
That way the unique selector is part of the t-call template and is not duplicated. It comes without any styling and the caller is responsible for the styling with a wrapping element. Then on layout change, it will just about moving the old div at the location of the new one. Not possible ?

rdeodoo replied:
> But in fact I am not sure it will be working since your use of small and p and ... etc. For example I wonder that if the user hits enter, it will create a new p which is probably not autorized in a small element and so might be extracted by smart browsers like Firefox ;)

That's not how the editor handles its basic elements (the ones in the topbar, `p`, `pre`, `small`, `h1` etc..). It does not add a `p` on enter but a new tag. So an enter inside a `small` will create another `small` to go next line.
Note that Shift Enter will simply enter `<br/>`.

Everything is worked perfectly fine on both Chrome and Firefox, not sure what bug are we referring to.
Did a quick video of the behavior, [see here](https://drive.google.com/file/d/12YFFYLtogbHyM_hJG4hm2rYbMM2GaAUR/view).

Am I missing the point or is there any valid reason to change current implementation about moving texts?

qsm-odoo replied:
I am not trying to make you change code for no reason @rdeodoo, I like the feature. I am just trying to make this a possible merge for today :wink: 

- Hit enter four times in discrete cookie bar, the `<p/>` is duplicated four times with "primary" class and nothing visually happens, go to classic, go back to discrete, not the same as before because the `<p/>` are converted to `<br/>` somehow.
- Hit enter in popup mode, you have two `<p/>` with "secondary" class. Go back to classic mode, you have one `<p/>` with one `<br/>` inside... and going back to popup mode will not solve the problem, so the semantic construction of your layout was broken by just trying another option.
- Like the guessed example I gave, add a table in your element, it pops out the element since `<p/>`/`<span/>` cannot contain a table and is lost beyond repair on layout switching. Same for using a list which seems nice to have in a cookie bar policy.
- ... etc

All of this cover the same point I tried to explain above: layout switching should just be about **one container node** (_to take into account everything that might have been done inside_) to **move** (_to not reconstruct DOM on layout change which may lead to undesired behavior like JS binding losses (and is also slower)_).

Bonus: Also I don't think using the `small` tag make sense here, it should be using the `small` class since you just wanna say it is smaller, not give any semantic meaning to the text if I am not mistaken? Unfortunately, the editor is not smart enough to remove a small class if the user desires to do it but it will be the case for 14.0 and at worst the user can use the font-size stuff.

Also use `o_` as classes prefix please :wink:

rdeodoo replied:
> Also use o_ as classes prefix please

:+1: 

 > Hit enter four times in discrete cookie bar, the `<p/>` is duplicated four times with "primary" class and nothing visually happens

This has nothing to do with my code (at least directly, indirectly I could've find another way to position stuff).
This is a limitation of css/editor. Normal enter is not working with `d-flex` and  `align-items-center`. It creates a new element (another `p`) but it is displayed next to the previous, so you visually don't see anything.
That something I wanted to discuss, see [here](https://github.com/odoo/odoo/compare/b3c95fa5c4ae470b9db080a5e325275b3cdcb581..3ced211957b0e5db35fd1edb97430cee1d7f7e0c#diff-24b0433302053618c5b742f3eded1d67L1063), but removed it from the PR at the end..

> Hit enter in popup mode, you have two `<p/>` with "secondary" class. Go back to classic mode, you have one `<p/>` with one `<br/>` inside... and going back to popup mode will not solve the problem, so the semantic construction of your layout was broken by just trying another option.
Like the guessed example I gave, add a table in your element, it pops out the element since `<p/>`/`<span/>` cannot contain a table and is lost beyond repair on layout switching. Same for using a list which seems nice to have in a cookie bar policy.

Ok, I got your point, everything of that I am aware of, purposely how I expected it to work.
Maybe bad design but that's what I chose to implement.
The goal of this was to preserve basic layouting as you can see [here](https://drive.google.com/file/d/12YFFYLtogbHyM_hJG4hm2rYbMM2GaAUR/view).

If we want to keep the whole stuff that was designed, I can do it. I think it will be overkill but whatever. I want discrete mode to remain discrete, just keep texts, not the whole layout that you constructed afterward in popup mode. Otherwise, when switching from the popup layout, that you fully customized, to discrete mode, it won't have much sense.
But I guess, if it bugged you, that 'losing' what you constructed when changing layout is not UX friendly.

Will try with your approach to keep everything.

qsm-odoo replied:
> I want discrete mode to remain discrete

Is not this the point of primary/secondary blocks ? You just keep primary if I remember well ? If not then maybe for discrete you can simply lose all text... but clearly not in preview mode :/ Just hovering an option should not make you lose everything.

>  I think it will be overkill but whatever.

The point is that doing it as I suggested results in **less and safer** code... and might prevent buggy behaviors (which are because of the editor, not your code... but your code runs inside the editor so we have to deal with it)... how is that overkill ?

rdeodoo replied:
> Is not this the point of primary/secondary blocks ? You just keep primary if I remember well ?

Not really, the goal is the have a nice sentence self sufficient in discrete mode that you can customize.
In other mode, that sentence is kept and used, but also another whole paragraph is added to have more content with more explanation on the cookies bar. When you change the primary sentence, the change is reflected in every mode.

> how is that overkill ?

-> _I want discrete mode to remain discrete, just keep texts, not the whole layout that you constructed afterward in popup mode. Otherwise, when switching from the popup layout, that you fully customized, to discrete mode, it won't have much sense._

Currently adapting the code anyway so this one can move forwards

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -1207,7 +1207,7 @@ var SnippetsMenu = Widget.extend({
             const $invisibleDOMPanelEl = $(this.invisibleDOMPanelEl);
             $invisibleDOMPanelEl.find('.o_we_invisible_entry').remove();
-            const $invisibleSnippets = this.getEditableArea().find('.o_snippet_invisible');
+            const $invisibleSnippets = this.getEditableArea().find('.o_snippet_invisible').addBack('.o_snippet_invisible');
```
Not needed anymore, I guess it was a bug induced by https://www.odoo.com/web#id=2122947&action=327&model=project.task&view_type=form&menu_id=4720 which is now reverted. But your fix will still be needed if the original problem of the mentioned task is solved another way. So maybe you should keep this but review your commit message :wink:

qsm-odoo commented on addons/website/static/src/snippets/s_popup/options.js:
```diff
@@ -41,7 +38,9 @@ options.registry.SnippetPopup = options.Class.extend({
         const $container = $(widgetValue === 'moveToFooter' ? 'footer' : 'main');
-        this.$target.closest('.s_popup').prependTo($container.find(':o_editable').first());
+        this.$target.closest('.s_popup').prependTo(
+            $container.find(':o_editable:not(.o_snippet_invisible):not(.o_snippet_invisible *)').first()
```
Same comment for this commit as the first one but here I would remove the commit, or adapt it here to `.oe_structure:o_editable` maybe ? And the option should be hidden if no such element exists ?

rdeodoo replied:
Actually, the revert commit you mentionned did not fix this one.
I fixed/avoided the issue by inserting the cookies template (inside instead of before div[id=footer]).
I squashed this one with the previous one and used `.oe_structure:o_editable` for a more generic safety nest.

qsm-odoo commented on addons/website/static/src/snippets/s_popup/000.js:
```diff
@@ -14,7 +15,9 @@ const PopupWidget = publicWidget.Widget.extend({
      */
     start: function () {
-        this._bindPopup();
+        if (!utils.get_cookie(this.$target.attr('id'))) {
```
Try not to use `$target` in widgets, it is meant to disappear some day. Use `this.$el` like for all widgets in Odoo.

qsm-odoo commented on addons/website/views/snippets/s_popup.xml:
```diff
@@ -57,6 +58,7 @@
                 <we-button data-select-data-attribute="mouseExit">On Exit</we-button>
             </we-select>
             <we-input class="o_we_inline" string="Delay" data-select-data-attribute="" data-attribute-name="showAfter" data-unit="s" data-save-unit="ms" data-dependencies="show_delay"/>
+            <we-input class="o_we_inline" string="Period between consents" data-select-data-attribute="7days" data-attribute-name="consentsDelay" data-unit="days"/>
```
Don't use data-unit for this, add days in a span next the the we-input I guess. Otherwise it won't be translated.

rdeodoo replied:
Fixed another way, I prefer to keep the DOM as it is supposed to be (we style).

qsm-odoo commented on addons/website/static/src/snippets/s_popup/000.js:
```diff
@@ -74,6 +77,8 @@ const PopupWidget = publicWidget.Widget.extend({
      * @private
      */
     _onCloseClick: function () {
+        const nbDays = this.$target.find('.s_popup_main').data('consentsDelay');
```
Sorry to annoy you with that but could you use `consentsDuration` instead ? Delay would mean "le temps que tu as pour consentir" to me :wink:

qsm-odoo commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,144 @@
+        <field name="key">website.cookie_policy</field>
+        <field name="arch" type="xml">
+            <t name="Cookie Policy" t-name="website.cookie_policy">
+                <t t-call="website.layout">
```
Remind me why this cannot be a `<template>` again?

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -168,6 +169,26 @@ def write(self, values):
+            else:
+                self.env['website.page'].search([
+                    ('website_id', '=', self.id),
+                    ('url', '=', '/cookie-policy'),
```
Should not it be searched by key ? What if the user changes the url ?

qsm-odoo commented on addons/website/static/src/js/backend/res_config_settings.js:
```diff
@@ -33,4 +44,36 @@ BaseSettingController.include({
+                close: true,
+                click: () => {
+                    return this._setValue(checked);
+                },
```
When your function is just a return, you can do `() => this._setValue(checked)` in one line. Same for cancelCallback ;)

qsm-odoo commented on addons/website/static/src/js/backend/res_config_settings.js:
```diff
@@ -1,7 +1,18 @@
+const QWeb = core.qweb;
+const _t = core._t;
+
+ajax.loadXML('/website/static/src/xml/website.res_config_settings.xml', QWeb);
```
Add that as xmlDependencies of your WebsiteCookiesbarField instead

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -3,9 +3,12 @@ odoo.define('website.editor.snippets.options', function (require) {
 var options = require('web_editor.snippets.options');
+// TODO (review) better way to do that? this shows `popupOptions` as unused
+//      but is required as we extend a registry Class defined in it..
+var popupOptions = require('website.s_popup_options');
```
Just `require` then, do not put the result in a variable. No need of comment either it is self explanatory: you don't need the result but you require it so that this module is called after.

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1027,6 +1030,61 @@ options.registry.anchor = options.Class.extend({
+options.registry.CookiesBar = options.registry.SnippetPopup.extend({
+    xmlDependencies: [
+        '/website/static/src/xml/website.cookies_bar.xml',
+    ],
```
One line :+1: SnippetPopup does not have xmlDependencies? Could be worth extending it anyway to be safe.

rdeodoo replied:
No, but better safe than sorry, addes the concat

qsm-odoo commented on addons/website/static/src/snippets/s_popup/000.scss:
```diff
@@ -2,6 +2,8 @@
     .s_popup_content {
         min-height: var(--o-we-toolbar-height);
+        // keep lower than <p> height (cookies bar)
+        min-height: $o-font-size-base * $o-line-height-base;
```
Choose one :)

qsm-odoo commented on addons/website/static/src/snippets/s_popup/options.js:
```diff
@@ -8,7 +8,7 @@ options.registry.SnippetPopup = options.Class.extend({
      */
     start: function () {
-        this.$target.find('.s_popup_close').on('click', () => {
+        this.$target.find('.js_close_popup').on('click', () => {
```
Why changing the `s_` one ?

rdeodoo commented:
Everything has been adapted.

sbuhl commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,137 @@
+                <div class="oe_structure">
+                    <section class="pt8 pb8">
+                        <div class="container">
+                            <h2 class="pt16">Cookies</h2>
```
Cookie Policy

sbuhl commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,137 @@
+                                Cookies are small bits of text sent by our servers to your computer or device when you access our services.
+                                They are stored in your browser and later sent back to our servers so that we can provide contextual content.
+                                Without cookies, using the web would be a much more frustrating experience.
+                                We use them to support your activities on our website, for example your session (so you don't have to login again) or your shopping cart.
```
We use them to support your activities on our website. For example, your session (so you don't have to login again) or your shopping cart.

sbuhl commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,137 @@
+                                better site experiences and tools in the future.
+                            </p>
+                            <p>
+                                We also use third-party services such as Google Analytics, who set and use their own cookies to identify visitors and provide their own contextual services.
```
Nécessaire cette phrase?

rdeodoo replied:
je dirais que non, vu qu'elle est 'répétée' plus bas. je la supprime

sbuhl commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,137 @@
+                                                    Used to make advertising more engaging to users and more valuable to publishers and advertisers,
+                                                    such as providing more relevant ads when you visit other websites that display ads or to improve reporting on ad campaign performance.
+                                                </p>
+                                                <p>Note that some third-party services may install additional cookies on your browser in order identify you.</p>
```
in order to identify you.

sbuhl commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,137 @@
+                                                </p>
+                                                <p>Note that some third-party services may install additional cookies on your browser in order identify you.</p>
+                                                <p>
+                                                    You may opt out of a third-party's use of cookies by visiting the <a href="https://optout.networkadvertising.org/?c=1" rel="nofollow">Network Advertising Initiative opt-out page</a>.
```
opt-out

Yenthe666 commented:
Just checking, will this land in the upcoming V14?

rdeodoo commented:
@Yenthe666 Hopefully :smile:

qsm-odoo commented on addons/website/data/website_data.xml:
```diff
@@ -115,6 +115,134 @@
+    <template id="cookie_policy" name="Cookie Policy">
+        <t t-call="website.layout">
+            <div id="wrap">
+                <div class="oe_structure">
```
Only one level

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1026,6 +1027,59 @@ options.registry.anchor = options.Class.extend({
+                this.allChildNodes[selector] = $el[0].childNodes;
+            }
+            $template.find(selector).html(this.allChildNodes[selector]);
+        }
```
I see you changed with the `<div and unique class> stuff` so you don't need to use the childNodes etc. Actually I am not sure what happens with `.html(childNodes)`, it is actually not allowed by jQuery documentation so I don't know if the nodes are recreated or not.

Can't the code be:
```javascript
        for (const selector of selectorsToKeep) {
            const $el = $content.find(selector);
            if ($el.length) {
                $template.find(selector).replaceWith($el);
            }
        }
```
No need of the tempory allChildNodes array right? + perform a move instead of a DOM duplication

rdeodoo replied:
No, it's not that trivial.
For layout A B C, some elements are only related to B and C for instance, we then need to keep the A-related element inside the array, for when the user will get back to A.
Also, some elements required in both A B C could be missing if the user deleted the element. eg in layout A, an element was removed, we need to restore it when he moves to B or C.
It goes both ways really, there might be missing element or not needed element when switching from a layout to another.

Anyway, despite the fact it was working as it should on both browser, as it was not jQuery compliant, I refactored that part to not pass DOM element to `$.html()`.

qsm-odoo replied:
Ok, I thought there might be something like it but I was not sure anymore. Note: it might not seem important to use append/replace, but I guarantee you this could be. It would be normal for an option / an editor code to rely on DOM nodes to not be duplicated (because saved in a variable / event handlers attached / whatever) :wink: 

Anyways, if you changed everything, let's send this to the testing you want. India? Sbu? :wink:

rdeodoo replied:
Task has already moved through the testing phase (sbu).
You can move forward with this one

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -1026,6 +1027,59 @@ options.registry.anchor = options.Class.extend({
+            $template.find(selector).html(this.allChildNodes[selector]);
+        }
+
+        $content.html($template);
```
Same comment: `$content.empty().append($template)`

qsm-odoo commented on addons/website/static/src/xml/website.res_config_settings.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates id="template" xml:space="preserve">
```
If you agree, I would unindent the first level of all your static files

rdeodoo replied:
Yes, I prefer that way too, it's just that lot of `/views/` files are doing it while most `/static/` don't so I kept the same stuff.

robodoo commented:
Merge method set to rebase and fast-forward

JKE-be commented:
:postal_horn: Apres un long combat de plus de 3 ans...

sbuhl commented:
Tellement d'émotion de la voir finalement mergée! 
Thanks everyone!

Yenthe666 commented:
Have a 🍪 to all of those involved. Great to see this added.
